### PR TITLE
Migrate to webpack

### PR DESCRIPTION
**Pull request by alexgibson**
## Description
- Migrate from Gulp to Webpack

## Issue / Bugzilla link
https://github.com/mozilla/bedrock/issues/10310

## Testing
Demo: https://www-demo2.allizom.org/en-US/

- To test using a local bedrock install make sure to run `yarn` before `npm start`.
- To test using the Docker install run `make build` before `make run`.

Successful test run: https://gitlab.com/mozmeao/www-config/-/pipelines/358631832

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -1,22 +1,35 @@
 module.exports = {
     env: {
-        'jasmine': true
+        'jasmine': true,
+        'commonjs': true,
```
Comment (pmac) [inline .eslintrc.js:4]: you can leave trailing comas now?!?! _hopes_

**Comment by pmac** [comment]
Looks like things are going great. It works for me locally. The home page cards aren't loading their images, which could be related the the contentful merge? I'm not sure yet. But looking very promising! Psyched!

**Comment by alexgibson** [comment]
> Looks like things are going great. It works for me locally. The home page cards aren't loading their images, which could be related the the contentful merge?

Thanks @pmac 👍 

The cards are loading ok for me, so my guess is you may need to update your master branch (this might be related to when the cards broke after merging the Contenful PR?).

**Comment by alexgibson** [comment]
@maureenlholland @craigcook @nathan-barrett @reemhamz r?

I think this is ready for review. I'd appreciate it if you coulkd all test the PR to make sure it's working for everyone before we merge. If you have any questions about it, I'd be happy to walk through things.

**Review by maureenlholland — COMMENTED** [review]
Using macOS Big Sur 11.4 and Apple M1 chip
docker test: Docker 3.6.0 (engine 20.10.8)
local test: Python 3.8.10

This works for me both with docker and local builds. I could change css and js files and see the browsersync reload. Send to device test worked as well. I like the switch to a resolved path instead of relative paths in scss files.

The only weirdness was a MODE warning despite the NODE_ENV setting in the npm command. I'm not sure what's going on here but it doesn't seem to be blocking anything.

File: media/js/ie/libs/html5shiv.js [hunk 2]
```diff
@@ -1,4 +1,4 @@
...
-m=/^<|^(?:a|b|button|code|div|fieldset|form|h1|h2|h3|h4|h5|h6|i|iframe|img|input|label|li|link|ol|option|p|param|q|script|select|span|strong|style|table|tbody|td|textarea|tfoot|th|thead|tr|ul)$/i,i,j;(function(){var a=b.createElement("a");a.innerHTML="<xyz></xyz>";i="hidden"in a;if(!(a=1==a.childNodes.length))a:{try{b.createElement("a")}catch(c){a=!0;break a}a=b.createDocumentFragment();a="undefined"==typeof a.cloneNode||"undefined"==typeof a.createDocumentFragment||"undefined"==typeof a.createElement}j=
-a})();var e={elements:d.elements||"abbr article aside audio bdi canvas data datalist details figcaption figure footer header hgroup main mark meter nav output progress section summary time video",shivCSS:!1!==d.shivCSS,shivMethods:!1!==d.shivMethods,type:"default",shivDocument:h};g.html5=e;h(b)})(this,document);
+/**
```
[the remark this answers is not in the document]
Comment (alexgibson) [reply media/js/ie/libs/html5shiv.js:1]: This is an updated version of the lib that works better with the output webpack produces

File: package.json [hunk 3]
```diff
@@ -34,24 +30,36 @@
...
     "lint": "npm run lint-js && npm run lint-css && npm run lint-json",
     "pretest": "npm run lint",
-    "test": "./node_modules/.bin/karma start ./tests/unit/karma.conf.js"
-  }
+    "test": "./node_modules/.bin/karma start ./tests/unit/karma.conf.js",
+    "static": "webpack --config webpack.static.config.js --bail",
+    "prebuild": "npm run static",
+    "build": "NODE_ENV=production webpack --bail",
+    "prewatch": "npm run static",
+    "watch": "NODE_ENV=development webpack --watch"
```
[the remark this answers is not in the document]
Comment (alexgibson) [reply package.json:59]: Interesting I'll take a look here, thanks for catching

File: docs/coding.rst [hunk 4]
```diff
@@ -8,6 +8,99 @@
 Developing on Bedrock
 =====================
 
+Asset Management and Bundling
```
[the remark this answers is not in the document]
Comment (alexgibson) [reply docs/coding.rst:11]: Good catch!

File: Dockerfile [hunk 5]
```diff
@@ -113,7 +112,8 @@ FROM app-base AS release
 
 RUN bin/run-sync-all.sh
 
-COPY --from=assets /app/static_final /app/static_final
+COPY --from=assets /app/assets /app/assets
```
[the remark this answers is not in the document]
Comment (alexgibson) [reply Dockerfile:115]: Just for naming sake really. We no longer need two separate build folders like we used to with Gulp, so the `final` part now seemed redundant :)

**Comment by alexgibson** [comment]
@maureenlholland I pushed a commit that should simplify how we set Webpack's build mode. Hopefully this should fix the warning you were seeing? I also removed the stray references to Gulp from the docs.

**Review by maureenlholland — COMMENTED** [review]
r+ 🕸️ ! Working great for me with no mode warning!

**Review by nathan-barrett — APPROVED** [review]
everything works great on my machine, both local installation and through docker. r+

File: Dockerfile [hunk 6]
```diff
@@ -11,14 +11,13 @@ COPY package.json yarn.lock ./
 
 # install dependencies
 RUN yarn install --pure-lockfile
-RUN yarn global add gulp-cli@2.2.1
 
 # copy supporting files and media
-COPY .eslintrc.js .eslintignore .stylelintrc .stylelintignore gulpfile.js ./
+COPY .eslintrc.js .eslintignore .stylelintrc webpack.config.js webpack.static.config.js ./
```
Comment (nathan-barrett) [inline Dockerfile:16]: do we no longer need a `.stylelintignore` ?

Comment (alexgibson) [reply Dockerfile:16]: oh, yes we do - good catch!

**Review by craigcook — APPROVED** [review]
This is all working great for me, with no errors.

**Review by reemhamz — APPROVED** [review]
Everything works as planned, not to mention much faster load time compared to Gulp (R.I.P.). This is great, Alex! 🎉

**Comment by pmac** [comment]
image

**Comment by alexgibson** [comment]
Thanks all!

I'm likely going to give this a final rebase and merge tomorrow (doing a deployment today to clear out the pipes). Exciting!

**Comment by alexgibson** [comment]
Latest passing test run https://gitlab.com/mozmeao/www-config/-/pipelines/371324994

**Pull request merged**